### PR TITLE
[fix]: [CDS-46422]: perform template version Validation For Remote Templates in get calls

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes
Added validation for versionLabel in get call, when template is remote.
      Validation happens on fields in request against yaml present on remote
Screenshot 2023-01-06 at 10 29 54 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)

### CODE REVIEW

**Review by utkarshchoubey-harness — COMMENTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 5 file(s) to review. 0 of 2 approvals obtained.

**Review by utkarshchoubey-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292#-:-NIlWLas1wJhFcxxS0Rp:bnfp4nl)* status: 5 file(s) to review. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
Lets say the validation fails for an existing remote template, user will start seeing an error while trying to load the template, I think we should wait for the UI change where we show the Yaml in case of these errors

**Comment by tchaurasiya** [comment]
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Lets say the validation fails for an existing remote template, user will start seeing an error while trying to load the template, I think we should wait for the UI change where we show the Yaml in case of these errors
</blockquote></details>

Make sense, But @yogesh-chauhan if lets say versionLabel(or any validated filed) is changed  on remote,
Will ui default yaml view point it out?

**Review by adivishy1 — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 5 file(s) to review. 1. unresolved discussions.

File: 840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java [hunk 1]
```diff
@@ -439,6 +447,22 @@
...
+    if (isNotBlank(templateEntity.getYaml())) {
+      YamlField templateYamlField = TemplateUtils.getTemplateYamlFieldElseThrow(templateEntity.getOrgIdentifier(),
+          templateEntity.getProjectIdentifier(), templateEntity.getIdentifier(), templateEntity.getYaml());
+
+      String templateVersionFromGit =
+          templateYamlField.getNode().getStringValue(YAMLFieldNameConstants.TEMPLATE_VERSION);
+      if (!(EmptyPredicate.isEmpty(versionLabel) && EmptyPredicate.isEmpty(templateVersionFromGit))
+          && !versionLabel.equals(templateVersionFromGit)) {
+        throw new InvalidRequestException(format(
+            "Template version from remote template file [%s] does not match with template version in request [%s]. Template version field is only supposed to change through harness application",
```
Comment (yogesh-chauhan) [inline 840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java:460]: Lets drop the "application" word. 
We should say
Each template version maps to a unique file on Git. Create a new version through harness or import a new version if the file is already created on Git

File: 840-template-service/src/main/java/io/harness/template/services/TemplateGitXServiceImpl.java [hunk 2]
```diff
@@ -163,26 +156,15 @@
 
   public void performImportFlowYamlValidations(String orgIdentifier, String projectIdentifier,
```
Comment (yogesh-chauhan) [inline 840-template-service/src/main/java/io/harness/template/services/TemplateGitXServiceImpl.java:157]: is this change needed now ?

File: 840-template-service/src/main/java/io/harness/template/utils/TemplateUtils.java [hunk 3]
```diff
@@ -7,15 +7,26 @@
...
+import io.harness.pms.yaml.YAMLFieldNameConstants;
+import io.harness.pms.yaml.YamlField;
+import io.harness.pms.yaml.YamlUtils;
 import io.harness.template.entity.TemplateEntity;
+import io.harness.yaml.validator.InvalidYamlException;
 
+import java.io.IOException;
+import java.util.Collections;
 import lombok.experimental.UtilityClass;
+import org.jetbrains.annotations.NotNull;
```
Comment (yogesh-chauhan) [inline 840-template-service/src/main/java/io/harness/template/utils/TemplateUtils.java:29]: lets not use NotNull from jetbrains

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Comment by tchaurasiya** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 450 at r4](https://reviewable.io/reviews/harness/harness-core/41292#-NIkxZAxx4g2pdbHkyU9-r4-450:-NLP34H57sqKOD2OcwI7:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a9536c129bec1d330a1e7df3d74d54ea19a433cd/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L450)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets add private methods at the end for better readability

Generally its good to follow


class {

 .. all public methods

 .. all private methods

}



</blockquote></details>

Done.

**Comment by tchaurasiya** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/TemplateGitXServiceImpl.java` line 157 at r4](https://reviewable.io/reviews/harness/harness-core/41292#-NIkxZAxx4g2pdbHkyUC-r4-157:-NLP36Vf1s_KnFWXpf2f:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a9536c129bec1d330a1e7df3d74d54ea19a433cd/840-template-service/src/main/java/io/harness/template/services/TemplateGitXServiceImpl.java#L157)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

is this change needed now ?
</blockquote></details>

Done.

**Comment by tchaurasiya** [comment]
*[`840-template-service/src/main/java/io/harness/template/utils/TemplateUtils.java` line 29 at r4](https://reviewable.io/reviews/harness/harness-core/41292#-NL6qjcHQrPgcbix-QVL-r4-29:-NLP37fS1Ftgao5_u5KB:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a9536c129bec1d330a1e7df3d74d54ea19a433cd/840-template-service/src/main/java/io/harness/template/utils/TemplateUtils.java#L29)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets not use NotNull from jetbrains
</blockquote></details>

Done.

**Review by tchaurasiya — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 6 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 460 at r4](https://reviewable.io/reviews/harness/harness-core/41292#-NIkxZAxx4g2pdbHkyU9-r4-460:-NLP35PI7ODIwfNyd3Rd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a9536c129bec1d330a1e7df3d74d54ea19a433cd/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L460)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

Lets drop the "application" word. 
We should say
Each template version maps to a unique file on Git. Create a new version through harness or import a new version if the file is already created on Git 
</blockquote></details>

Done.

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r3, 1 of 4 files at r4, 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1253 at r5](https://reviewable.io/reviews/harness/harness-core/41292#-NLQaC_F5oKAhcVwArOQ:-NLQ_iY6AllI7D7OGZiw:bb3r7u1) ([raw file](https://github.com/harness/harness-core/blob/a19fd1f133cc945c1f839ebd63537073b5cea02f/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1253)):*
> ```Java
>       String templateVersionFromGit =
>           templateYamlField.getNode().getStringValue(YAMLFieldNameConstants.TEMPLATE_VERSION);
>       if (!(EmptyPredicate.isEmpty(versionLabel) && EmptyPredicate.isEmpty(templateVersionFromGit))
> ```

lets simplify for readability

**Comment by tchaurasiya** [comment]
*[`840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java` line 1253 at r5](https://reviewable.io/reviews/harness/harness-core/41292#-NLQaC_F5oKAhcVwArOQ:-NLRMshRA5uAa4v5MqUG:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a19fd1f133cc945c1f839ebd63537073b5cea02f/840-template-service/src/main/java/io/harness/template/services/NGTemplateServiceImpl.java#L1253)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

lets simplify for readability
</blockquote></details>

Done.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
trigger messagecheck

**Review by adivishy1 — APPROVED** [review]
> Reviewed 1 of 4 files at r4, 4 of 5 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41292)* status: :shipit: complete! ✓

**Pull request merged**